Assets/_Challenges/Scripts/AI Behaviours/Debug/LogBehaviour.cs

```diff
@@ -11,15 +12,46 @@ namespace MessyCoderCommunity.AI
     [CreateAssetMenu(fileName = "Log Behaviour", menuName = "Messy AI/Debug/Log")]
     public class LogBehaviour : GenericAiBehaviour<GameObject>
     {
-        [Tooltip("The message to display. This can include variables on the chalkboard using `{VARIABLE_NAME}`.")]
-        public string message = "{agent} says 'Hi'";
+        [SerializeField, Tooltip("The message to display. This can include variables on the chalkboard using `{VARIABLE_NAME}`.")]
+        string message = "{agent} says 'Hi'";
+
+        private Regex variableRegex;
+
+        public override void Initialize(GameObject agent, Chalkboard chalkboard)
+        {
+            base.Initialize(agent, chalkboard);
+
+            variableRegex = new Regex(@"\{([^}]*)\}");
+        }
 
         public override void Tick(Chalkboard chalkboard)
         {
+            string expandedMessage = message;
+            MatchCollection matches = variableRegex.Matches(expandedMessage);
+            for (int i = 0; i < matches.Count; i ++)
+            {
+                string token = matches[i].Groups[0].Value;
+                string variableName = matches[i].Groups[1].Value;
+
+                UnityEngine.Object unityValue = chalkboard.GetUnity<UnityEngine.Object>(variableName.GetHashCode());
+                if (unityValue != null)
+                {
+                    expandedMessage = expandedMessage.Replace(token, unityValue.ToString());
+                } else
+                {
+                    System.Object systemValue = chalkboard.GetSystem<System.Object>(variableName.GetHashCode());
+                    if (systemValue != null)
+                    {
+                        expandedMessage = expandedMessage.Replace(token, systemValue.ToString());
```

Comment:
We might want to use position based reconstruction rather than replace if this is in a hot path.

Reply:
I assume this is for performance reasons? If the RE.match provides the position then I think this makes sense. Otherwise I don't know that it is worth the refactor, this is only used when debugging. Maybe wrap the whole behaviour with an `if UNITY_EDITOR"?

Reply:
I took a look and match does provide the index of the match. However, there seems to be some debate about the most efficient way to replace substrings. I'm just going to pick the most readable since this is a debug method and thus the odd ms is not an issue.